pcl_detector/src/pcl_detector_ros.cpp

```diff
@@ -0,0 +1,143 @@
+  on_set_callback_handle_ = add_on_set_parameters_callback(std::bind(&PclDetectorNode::parametersCallback, this, std::placeholders::_1));
+  
+  // Initialize the detector with configured parameters
+  std::string detector;
+  get_parameter<std::string>("detector", detector);
+  m_detector = initialize_detector(detector);
+}
+
+// Callback function for parameter changes
+rcl_interfaces::msg::SetParametersResult PclDetectorNode::parametersCallback(const std::vector<rclcpp::Parameter> &parameters) {
+    std::string detector_name;
+    rcl_interfaces::msg::SetParametersResult result;
+
+    for (const auto &parameter : parameters) {
+        if (parameter.get_name() == "detector") {
+            if (detector_type.count(parameter.as_string()) == 0) {
+                result.successful = false;
+                return result;
+            }
+        }
+    }
+    
+    parameters_changed = true;
+    result.successful = true;
+    result.reason = "success";
+    return result;
+}
+
+// Initialize the selected detector based on the parameter
+std::unique_ptr<IPclDetector> PclDetectorNode::initialize_detector(std::string detector)
+{
+    std::unique_ptr<IPclDetector> configured_detector = nullptr;
+
+    switch (detector_type[detector]) {
+        case DetectorType::Euclidean: {
+            float cluster_tolerance;
+            get_parameter<float>("euclidean.cluster_tolerance", cluster_tolerance);
+            int min_points;
+            get_parameter<int>("euclidean.min_points", min_points);
+            configured_detector = std::make_unique<EuclideanClusteringDetector>(cluster_tolerance, min_points);
+            RCLCPP_INFO_STREAM(this->get_logger(), "Euclidean detector initialized with cluster tolerance = " << cluster_tolerance << " and min_points = " << min_points);
+            break;
+        }
+
+        case DetectorType::DBSCAN: {
+            float eps; 
+            get_parameter<float>("dbscan.epsilon", eps);
+            int min_points;
+            get_parameter<int>("dbscan.min_points", min_points);
+            configured_detector = std::make_unique<DBSCANDetector>(eps, min_points);
+            RCLCPP_INFO_STREAM(this->get_logger(), "DBSCAN detector initialized with eps = " << eps << " and min_points = " << min_points);
+            break;
+        }
+
+        default: {
+            RCLCPP_ERROR(this->get_logger(), "Invalid pcl detector specified! Shutting down...");
+            rclcpp::shutdown();
+        }
+    }
+
+    return configured_detector;
+}
+
+// Callback function for processing incoming point cloud messages
+void PclDetectorNode::topic_callback(const sensor_msgs::msg::PointCloud2::SharedPtr cloud_msg)
+{
+    // Check if parameters have changed
+    if (parameters_changed) {
+        std::string detector;
+        get_parameter<std::string>("detector", detector);
+        
+        m_detector = initialize_detector(detector);
+        parameters_changed = false;
+    }
```

Comment:
This is a scuffed way to do this as we discussed @jorgenfj and @etfroeland. The problem @chrstrom is that you can't use the new parameters to set stuff from within a parameter callback function. ROS2 Iron adds a "post-parameter-update-callback" that lets us use the new parameters, but we can't use that. Can you think of a better solution?

Reply:
I'm not sure if the original functionality of enabling rapid switching between detector types like this should be supported, as it was mainly a tool for debugging/testing. I'm not aware of an elegant way of doing this either atm...

What is the status for dynamic-reconfigure in ROS2?